demos/demo_non_Gauss_Likelihood.py

```diff
@@ -0,0 +1,63 @@
+# %%
+import torch as xp
+import cuqi
+
+from cuqipy_pytorch.distribution import Cauchy, Laplace, Gaussian
+from cuqipy_pytorch.model import add_to_autograd
+from cuqipy_pytorch.sampler import NUTS
+
+# %% Forward models
+n = 50
+
+# Testproblem
+TP = cuqi.testproblem.Deconvolution1D(dim=n)
+
+# CUQIpy model added to autograd framework. y = Ax
+A = add_to_autograd(TP.model)
+y_obs = xp.as_tensor(TP.data)
+
+# Lambda function as forward model. d = Bu
+B = lambda u: xp.log(u**2 + 1) + xp.sin(u) # Contrived example
+
+# %% Bayesian models (non-Gaussian likelihoods)
+
+# Model 1
+x = Gaussian(xp.zeros(n), 0.1)                  # x ~ N(0, 1)
+y = Cauchy(location=A(x), scale=0.1)               # y ~ Cauchy(Ax, 0.1)
+J1 = cuqi.distribution.JointDistribution(x, y)
+
+# Model 2
+u = Gaussian(xp.zeros(n), 1)                    # u ~ N(0, 1)
+d = Laplace(location=B, scale=0.1*xp.ones(n))   # d ~ Laplace(Bu, 0.1)
+J2 = cuqi.distribution.JointDistribution(u, d)
+
+# %% Posteriors
+P1 = J1(y=y_obs)
+print(P1)
+
+P2 = J2(d=xp.ones(n))
+print(P2)
+
+# %% Evaluating logpdf's
+print(P1.logpdf(x=xp.ones(n)))
+print(P2.logpdf(u=xp.ones(n)))
+
+# %% Sampling
+
+# A convenience function to sample a Bayesian model
+# It creates the joint and posterior distributions as above.
+def sample_posterior(*densities, n_samples=500, n_burnin=500, **data):
+    """ Sample posterior given by a list of densities. The observations are given as keyword arguments. """
+    P = cuqi.distribution.JointDistribution(*densities)
```

Comment:
It seems to me that creating the Joint distribution is happening twice in this demo. Can we avoid this or explain briefly why we do it twice, if for illustration purpose. (e.g. line 32 and line 51)